src/client.cpp

```diff
@@ -1386,3 +1459,158 @@ int CClient::EstimatedOverallDelay ( const int iPingTimeMs )
+
+    iActiveChannels -= 1;
+
+    qInfo() << qUtf8Printable ( QString ( "> Freed client ch %1 for server ch %2; chan count = %3" )
+                                    .arg ( iClientChannelID )
+                                    .arg ( iServerChannelID )
+                                    .arg ( iActiveChannels ) );
+}
+
+// find, and optionally create, a client channel for the supplied server channel ID
+// returns a client channel ID or INVALID_INDEX
+int CClient::FindClientChannel ( const int iServerChannelID, const bool bCreateIfNew )
+{
+    QMutexLocker locker ( &MutexChannels );
+
+    if ( iServerChannelID == INVALID_INDEX || iServerChannelID >= MAX_NUM_CHANNELS )
+    {
+        return INVALID_INDEX;
+    }
+
+    int iClientChannelID = clientChannelIDs[iServerChannelID];
+
+    if ( iClientChannelID != INVALID_INDEX )
+    {
+        Q_ASSERT ( clientChannels[iClientChannelID].iServerChannelID == iServerChannelID );
+
+        return iClientChannelID;
+    }
+
+    // no matching client channel - create new one if requested
+    if ( bCreateIfNew )
+    {
+        // search clientChannels[] for a free client channel
+        for ( iClientChannelID = 0; iClientChannelID < MAX_NUM_CHANNELS; iClientChannelID++ )
+        {
+            CClientChannel* clientChan = &clientChannels[iClientChannelID];
+
+            if ( clientChan->iServerChannelID == INVALID_INDEX )
+            {
+                clientChan->iServerChannelID = iServerChannelID;
+                clientChan->iJoinSequence    = ++iJoinSequence;
+
+                clientChan->oldGain = clientChan->newGain = 1.0f;
+
+                clientChan->level = 0;
+
+                clientChannelIDs[iServerChannelID] = iClientChannelID;
+
+                iActiveChannels += 1;
+
+                qInfo() << qUtf8Printable ( QString ( "> Alloc client ch %1 for server ch %2; chan count = %3" )
+                                                .arg ( iClientChannelID )
+                                                .arg ( iServerChannelID )
+                                                .arg ( iActiveChannels ) );
+
+                return iClientChannelID; // new client channel ID
+            }
+        }
+    }
+
+    return INVALID_INDEX;
+}
+
+bool CClient::ReorderLevelList ( CVector<uint16_t>& vecLevelList )
+{
+    QMutexLocker locker ( &MutexChannels );
+
+    // vecLevelList is sent from server ordered by server channel ID
+    // re-order it by client channel ID before passing up to the GUI
+    // the list is passed in by reference and modified in situ
+
+    // check it is the right length
+    if ( vecLevelList.Size() != iActiveChannels )
+    {
+        return false; // tell caller to ignore it
+    }
+
+    int iClientCh;
+    int iServerCh = 0;
+
+    // fetch levels by server channel ID
+
+    for ( int i = 0; i < iActiveChannels; i++ )
+    {
+        // find next active server channel
+        while ( iServerCh < MAX_NUM_CHANNELS )
```

Comment:
Maybe we could find all "holes" first and then fill them one by one.

Reply:
I've put a lot of thought into this. The level list is received in ascending order of server channel. So when processing it, we first need a single pass through the list, skipping unused server channels, and placing the level value in the corresponding client channel. The we go through the client channels in order, fetching the level values from the active channels and placing them back in the list in order, so it can be passed up to the mixer board in the order it expects. So just two linear passes, and only as far as the number of active channels, and I can't think of any way that could be made more efficient. In practice we only need to skip a few channels, and never need to go all the way up to MAX_NUM_CHANNELS, as we will run out of list entries as soon as we have done all the active channels.

Reply:
Bear in mind that we are not resetting `iServerCh` each time round the outer loop. As the levels are in channel order, we just keep going from where we left off, looking for the next active channel, which will probably be next, or only a small number of steps away.